e2e topology manager: single-numa-node multi container tests

ffromani opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This is a followup of https://github.com/kubernetes/kubernetes/pull/87645 . I'm working to add more e2e tests for the topology manager
to fix the issue https://github.com/kubernetes/kubernetes/issues/83481
This PR adds minor fixes and improvements that didn't make it into https://github.com/kubernetes/kubernetes/pull/87645 and adds e2e node tests for multi-container pods.

**Which issue(s) this PR fixes**:
Related to https://github.com/kubernetes/kubernetes/issues/83481

**Special notes for your reviewer**:
N/A

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

ffromani committed:
e2e: topomgr: validate all containers in pod

Up until now, the test validated the alignment of resources
only in the first container in a pod. That was just an overlook.
With this patch, we validate all the containers in a given pod.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: topomgr: autodetect NUMA position of VF devs

Add autodetection code to figure out on which NUMA node are
the devices attached to.
This autodetection work under the assumption all the VFs in
the system must be used for the tests.
Should not this be the case, or in general to handle non-trivial
configurations, we keep the annotations mechanism added to the
SRIOV device plugin config map.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: topomgr: early check to detect VFs, not PFs

The e2e_node topology_manager check have a early, quick check
to rule out systems without sriov device, thus skipping the tests.

The first version of the ckeck detected PFs, (Physical Functions),
under the assumption that VFs (Virtual Functions) were already been
created. This works because, obviously, you can't have VFs without PFs.

However, it's a little safer and easier to understand if we check
firectly for VFs, bailing out from systems which don't provide them.

Nothing changes for properly configured test systems.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: topomgr: remove single-numa node hack

On single-NUMA node systems the numa_node of sriov devices was
sometimes reported as "-1" instead of, say, 0. This makes some
tests that should succeed[0] fail unexpectedly.

The reporting works as expected on real multi-NUMA node systems.

This small workaround was added to handle this corner case,
but it makes overall the code less readable and a bit too lenient,
hence we remove it.

+++

[0] on a single NUMA node system some resources are obviously
always aligned if the pod can be admitted. It boils down to the
node capacity at pod admittal time.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: topomgr: add multi-container tests

Add tests to check alignment of pods which contains more than one
container.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/cc @vpickard

fromanirh commented:
/sig node

fromanirh commented:
/assign @nolancon

fromanirh commented:
/hold
The PR is reviewable from my perspective but the final round of testing is still in progress

fromanirh commented:
/assign @derekwaynecarr

fromanirh commented:
bad rebase, missing one patch which causes a false negative. Fixing ASAP.

fromanirh commented:
/hold cancel
all tests succesful on my side

ffromani committed:
e2e: topomgr: extend tests to all the policies

Per https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/0035-20190130-topology-manager.md#multi-numa-systems-tests
we validate only the results for single-numa node policy,
because the is no a simple and reliable way to validate
the allocation performed by the other policies.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: topomgr: properly clean up after completion

Due to an oversight, the e2e topology manager tests
were leaking a configmap and a serviceaccount.
This patch ensures a proper cleanup

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
i signed it

nolancon commented:
/ok-to-test

fromanirh commented:
/retest

fromanirh commented:
force-pushed the series because the -verify job highlighted some issues with patches early in the series.

cynepco3hahue commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -598,7 +627,16 @@ func getSRIOVDevicePluginConfigMap(cmFile string) *v1.ConfigMap {
+	resourceAmount int64
+}
+
+func setupSRIOVConfigOrFail(f *framework.Framework, configMap *v1.ConfigMap) sriovData {
```
let's avoid the copy of the struct so return *sriovData

cynepco3hahue commented on test/e2e_node/numa_alignment.go:
```diff
@@ -184,22 +177,35 @@ func makeEnvMap(logs string) (map[string]string, error) {
+	sriovResourceName string
+}
+
+func containerWantsDevices(cnt *v1.Container, envInfo testEnvInfo) bool {
```
pass the pointer on `testEnvInfo`, the same true for all appearances of `testEnvInfo`

cynepco3hahue commented on test/e2e_node/numa_alignment.go:
```diff
@@ -184,22 +177,35 @@ func makeEnvMap(logs string) (map[string]string, error) {
+	return found
+}
+
+func checkNUMAAlignment(f *framework.Framework, pod *v1.Pod, cnt *v1.Container, logs string, envInfo testEnvInfo) (numaPodResources, error) {
```
return pointer on `numaPodResources`, the same true for all appearances of `numaPodResources`
and you can return `return nil, err`

cynepco3hahue commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -255,25 +297,17 @@ func findSRIOVResource(node *v1.Node) (string, int64) {
-	err := f.ClientSet.CoreV1().Pods(namespace).Delete(context.TODO(), name, &deleteOptions)
-	framework.ExpectNoError(err)
-}
+func validatePodAlignment(f *framework.Framework, pod *v1.Pod, envInfo testEnvInfo) {
```
pass pointer on `testEnvInfo`

cynepco3hahue commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -502,21 +536,27 @@ func runTopologyManagerPolicySuiteTests(f *framework.Framework) {
+	}, 2*time.Minute, 1*time.Second).Should(gomega.BeTrue())
+}
+
+func runTopologyManagerPositiveTest(f *framework.Framework, numPods int, ctnAttrs []tmCtnAttribute, envInfo testEnvInfo) {
```
pass pointer on `testEnvInfo`

cynepco3hahue commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -632,14 +670,34 @@ func setupSRIOVConfigOrFail(f *framework.Framework, configMap *v1.ConfigMap) (*v
-		dpPod.Namespace, dpPod.Name, dpPod.Spec.Containers[0].Name)
-	deletePodInNamespace(f, dpPod.Namespace, dpPod.Name)
-	waitForContainerRemoval(dpPod.Spec.Containers[0].Name, dpPod.Name, dpPod.Namespace)
+func teardownSRIOVConfigOrFail(f *framework.Framework, sd sriovData) {
```
pass pointer on `sriovData`

cynepco3hahue commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -648,43 +706,173 @@ func runTopologyManagerNodeAlignmentSuiteTests(f *framework.Framework, configMap
+		ctnAttrs = []tmCtnAttribute{
+			{
+				ctnName:       "gu-container",
+				cpuRequest:    "1000m",
```
Just curious, does CPU manager returns correct topology hints when you do not specify integer CPU's, I was sure it works only for static policy.

fromanirh replied:
In the `configureTopologyManagerInKubelet` helper we always set the CPUManager policy to `static` indeed.

cynepco3hahue replied:
Thanks for the clarification.

ffromani committed:
e2e: topomgr: address reviewer comments

Signed-off-by: Francesco Romani <fromani@redhat.com>

cynepco3hahue commented:
lgtm

derekwaynecarr commented:
the tests look good, appreciate the coverage with both positive and negative scenarios.

/lgtm
/approve
